autoload/targets.vim

```diff
@@ -39,16 +62,25 @@ function! targets#e(mapmode, modifier, original)
         return a:original
     endif
 
-    let char1 = nr2char(getchar())
-    let [trigger, which, chars] = [char1, 'c', char1]
-    for i in range(2)
-        if g:targets_nl[i] ==# trigger
-            " trigger was which, get another char for trigger
-            let char2 = nr2char(getchar())
-            let [trigger, which, chars] = [char2, 'nl'[i], chars . char2]
-            break
-        endif
-    endfor
+    let [nKeys, lKeys] = targets#getKeysAsList(g:targets_nl)
+    let keys = [s:getKeyAsStr()]
+    let chars = join(keys, '')
+    while (len(chars) <= len(nKeys) && s:hasPrefix(nKeys, chars)) || 
+          \ (len(chars) <= len(lKeys) && s:hasPrefix(lKeys, chars))
+        let nextKey = s:getKeyAsStr()
+        let keys = keys + [nextKey]
+        let chars = chars . nextKey
+    endwhile
```

Comment:
It took we quite a while to notice that we not only collect the n/l keys, but also the trigger. I played around with the code a bit and came up with this form.
```vim
let pending = ''
while 1
    let pending .= nr2char(getchar())

    if pending == nKeys
        let which = 'n'
        let trigger = nr2char(getchar())
        let typed = a:original . pending. trigger
        break
    endif

    if pending == lKeys
        let which = 'l'
        let trigger = nr2char(getchar())
        let typed = a:original . pending. trigger
        break
    endif

    if s:hasPrefix(nKeys, pending)
        continue
    endif

    if s:hasPrefix(lKeys, pending)
        continue
    endif

    if len(pending) > 1
        return a:original . pending
    endif

    let which = 'c'
    let trigger = pending
    let typed = a:original . pending
    break
endwhile
```
I think this one could also be extended to allow multi character triggers.
(by adding another break if pending equals a trigger and another continue if pending is prefix of a trigger)
(I don't think we should do that in this PR though)

What do you think?

Reply:
> It took we quite a while to notice that we not only collect the n/l keys, but also the trigger. 
Is this an issue?
> I played around with the code a bit and came up with this form.
> 
> ```viml
> let pending = ''
> while 1
>     let pending .= nr2char(getchar())
> 
>     if pending == nKeys
>         let which = 'n'
>         let trigger = nr2char(getchar())
>         let typed = a:original . pending. trigger
>         break
>     endif
> 
>     if pending == lKeys
>         let which = 'l'
>         let trigger = nr2char(getchar())
>         let typed = a:original . pending. trigger
>         break
>     endif
> 
>     if s:hasPrefix(nKeys, pending)
>         continue
>     endif
> 
>     if s:hasPrefix(lKeys, pending)
>         continue
>     endif
> 
>     if len(pending) > 1
>         return a:original . pending
>     endif
> 
>     let which = 'c'
>     let trigger = pending
>     let typed = a:original . pending
>     break
> endwhile
> ```
> 
> I think this one could also be extended to allow multi character triggers.
> (by adding another break if pending equals a trigger and another continue if pending is prefix of a trigger)
> (I don't think we should do that in this PR though)
> 
> What do you think?
I like it and replaced the previous implementation with it (with slight modifications). Please take another look to make sure I didn't mess anything up.